src/shims/tls.rs

```diff
@@ -217,16 +310,53 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 StackPopCleanup::None { cleanup: true },
             )?;
 
-            // step until out of stackframes
-            this.run()?;
+            this.enable_thread(active_thread)?;
+            return Ok(true);
+        }
+        this.machine.tls.dtors_running.get_mut(&active_thread).unwrap().last_dtor_key = None;
+
+        Ok(false)
+    }
+}
+
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+
+    /// Schedule an active thread's TLS destructor to run on the active thread.
+    /// Note that this function does not run the destructors itself, it just
+    /// schedules them one by one each time it is called and reenables the
+    /// thread so that it can be executed normally by the main execution loop.
+    ///
+    /// FIXME: we do not support yet deallocation of thread local statics.
+    /// Issue: https://github.com/rust-lang/miri/issues/1369
+    fn schedule_next_tls_dtor_for_active_thread(&mut self) -> InterpResult<'tcx> {
```

Comment:
Maybe also add a link to https://github.com/rust-lang/rust/issues/28129: we treat all threads equally here, and *do* run dtors on the main thread, as that seems most consistent -- but it is unclear if that matches actual behavior. But if we don't do that Miri reports memory leaks.